### PR TITLE
Use multiple threads while reading the metadata of tables to restore

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Improvement

### Changelog entry:
Use multiple threads while reading the metadata of tables from a backup while executing the RESTORE command.

Before this PR the metadata of tables was always read from a backup in a single thread. That was slow if the number of tables in a backup was high.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 1eafcdaf8768f252135f71a239c5ba561e6dfa87 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60040/1eafcdaf8768f252135f71a239c5ba561e6dfa87/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60040/1eafcdaf8768f252135f71a239c5ba561e6dfa87/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by vitlibar** [comment]
@SmitaRKulkarni Please take a look

File: src/Backups/RestorerFromBackup.h [hunk 1]
```diff
@@ -77,6 +80,7 @@ class RestorerFromBackup : private boost::noncopyable
     BackupPtr backup;
     ContextMutablePtr context;
     QueryStatusPtr process_list_element;
+    std::function<void()> after_each_task_callback;
```
Comment (SmitaRKulkarni) [inline src/Backups/RestorerFromBackup.h:83]: maybe we can call it `after_task_callback`, what do you think ?

**Comment by SmitaRKulkarni** [comment]
@vitlibar 
Do you think it makes sense to wrap this under a flag ?

**Review by SmitaRKulkarni — APPROVED** [review]
Rest all LGTM

File: src/Backups/RestorerFromBackup.h (hunk 1 above)
Comment (vitlibar) [reply src/Backups/RestorerFromBackup.h]: I've renamed it.

File: src/Backups/RestorerFromBackup.cpp [hunk 2]
```diff
@@ -828,45 +1007,47 @@ void RestorerFromBackup::insertDataToTable(const QualifiedTableName & table_name
 
 void RestorerFromBackup::addDataRestoreTask(DataRestoreTask && new_task)
 {
-    if (current_stage == Stage::INSERTING_DATA_TO_TABLES)
+    if (current_stage != Stage::INSERTING_DATA_TO_TABLES)
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/Backups/RestorerFromBackup.cpp:1010]: That's not a typo. [Earlier](https://github.com/ClickHouse/ClickHouse/blob/f8ff15a023fcfc8dc85b75658901c30a77a9d5a6/src/Backups/RestorerFromBackup.cpp#L136) the stages were like this:
```
CREATING_TABLES stage:
    - create tables
    - make data restore tasks to fill those tables with data (storage->restoreDataFromBackup() is called for each table)
INSERTING_DATA_TO_TABLES stage:
    - run data restore tasks in parallel (no data restore tasks could be added at this point)
```

Now in this PR I've changed that, so now it looks as following:
```
CREATING_TABLES stage:
    - create tables
INSERTING_DATA_TO_TABLES stage:
    - make data restore tasks to fill those tables with data (storage->restoreDataFromBackup() is called for each table in parallel with other tables)
    - run data restore tasks in parallel (other data restore tasks can be added at this point)
```

So now it's perfectly ok to add data restore tasks on the `INSERTING_DATA_TO_TABLES` stage.

**Comment by vitlibar** [comment]
> @vitlibar Do you think it makes sense to wrap this under a flag ?

I don't think we should provide user a choice here, because using multiple threads seems to be faster and as correct as it was.

**Comment by vitlibar** [comment]
Test failures:
- [02122_parallel_formatting_Template](https://s3.amazonaws.com/clickhouse-test-reports/60040/0df2a645cdaee3b6204fa802f65e945d5b747f0b/stateless_tests__msan__[2_6].html) - known, see [issue 58829](https://github.com/ClickHouse/ClickHouse/issues/58829)
- [02962_system_sync_replica_lightweight_from_modifier](https://s3.amazonaws.com/clickhouse-test-reports/60040/0df2a645cdaee3b6204fa802f65e945d5b747f0b/stateless_tests__tsan__[5_5].html) - see [issue 60480](https://github.com/ClickHouse/ClickHouse/issues/60480)
- [test_failed_mutations/test.py::test_exponential_backoff_with_merge_tree[node0-True]](https://s3.amazonaws.com/clickhouse-test-reports/60040/0df2a645cdaee3b6204fa802f65e945d5b747f0b/integration_tests__tsan__[2_6].html) - known, see [PR 58036](https://github.com/ClickHouse/ClickHouse/pull/58036#issuecomment-1965612326)

**Pull request merged**